File: flut/.flutter/lib/flutter/widgets/focus_traversal.dart
import 'package:flutter/material.dart';
import 'package:flut/flut/object.dart';
import 'package:flut/flut/runtime.dart';

class FlutTraversalDirection extends FlutEnumObject<TraversalDirection> {
  const FlutTraversalDirection()
    : super('TraversalDirection', const {
        'up': TraversalDirection.up,
        'right': TraversalDirection.right,
        'down': TraversalDirection.down,
        'left': TraversalDirection.left,
      });
}

class FlutTraversalEdgeBehavior extends FlutEnumObject<TraversalEdgeBehavior> {
  const FlutTraversalEdgeBehavior()
    : super('TraversalEdgeBehavior', const {
        'closedLoop': TraversalEdgeBehavior.closedLoop,
        'leaveFlutterView': TraversalEdgeBehavior.leaveFlutterView,
        'parentScope': TraversalEdgeBehavior.parentScope,
        'stop': TraversalEdgeBehavior.stop,
      });
}

class FlutFocusTraversalPolicy extends FocusTraversalPolicy
    with FlutAbstractObject {
  @override
  final FlutRuntime runtime;
  final int? sortDescendantsId;
  final int? invalidateScopeDataId;
  final int? changedScopeId;

  FlutFocusTraversalPolicy._(
    this.runtime,
    this.sortDescendantsId,
    this.invalidateScopeDataId,
    this.changedScopeId,
  );

  @override
  Iterable<FocusNode> sortDescendants(
    Iterable<FocusNode> descendants,
    FocusNode currentNode,
  ) {
    if (sortDescendantsId == null) return descendants;
    final result = runtime.callAction<List<dynamic>>(
      sortDescendantsId!,
      args: [descendants.toList(), currentNode],
    );
    if (result == null) return descendants;
    return result
        .map((item) => runtime.decodeObject<FocusNode>(item))
        .nonNulls
        .toList();
  }

  @override
  void invalidateScopeData(FocusScopeNode node) {
    super.invalidateScopeData(node);
    if (invalidateScopeDataId == null) return;
    runtime.callAction(invalidateScopeDataId!, args: [node]);
  }

  @override
  void changedScope({FocusNode? node, FocusScopeNode? oldScope}) {
    super.changedScope(node: node, oldScope: oldScope);
    if (changedScopeId == null) return;
    runtime.callAction(changedScopeId!, args: [node, oldScope]);
  }

  @override
  FocusNode? findFirstFocusInDirection(
    FocusNode currentNode,
    TraversalDirection direction,
  ) {
    throw UnimplementedError(
      'findFirstFocusInDirection is not supported in custom Python traversal policies',
    );
  }

  @override
  bool inDirection(FocusNode currentNode, TraversalDirection direction) {
    throw UnimplementedError(
      'inDirection is not supported in custom Python traversal policies',
    );
  }

  static FocusTraversalPolicy? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return FlutFocusTraversalPolicy._(
      runtime,
      runtime.unpackOptionalField<int>(data, 'sortDescendants'),
      runtime.unpackOptionalField<int>(data, 'invalidateScopeData'),
      runtime.unpackOptionalField<int>(data, 'changedScope'),
    );
  }
}

class FlutReadingOrderTraversalPolicy {
  FlutReadingOrderTraversalPolicy._();

  static ReadingOrderTraversalPolicy? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return ReadingOrderTraversalPolicy();
  }
}

class FlutOrderedTraversalPolicy {
  FlutOrderedTraversalPolicy._();

  static OrderedTraversalPolicy? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return OrderedTraversalPolicy(
      secondary: runtime.unpackOptionalField<FocusTraversalPolicy>(
        data,
        'secondary',
      ),
    );
  }
}

class FlutWidgetOrderTraversalPolicy {
  FlutWidgetOrderTraversalPolicy._();

  static WidgetOrderTraversalPolicy? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return WidgetOrderTraversalPolicy();
  }
}

class FlutNumericFocusOrder extends FlutValueObject {
  final NumericFocusOrder order;
  const FlutNumericFocusOrder(this.order) : super('NumericFocusOrder');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    result['order'] = order.order;
    return result;
  }

  static NumericFocusOrder? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return NumericFocusOrder(
      runtime.unpackRequiredField<double>(data, 'order'),
    );
  }
}

class FlutFocusTraversalGroup {
  FlutFocusTraversalGroup._();

  static FocusTraversalGroup? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return FocusTraversalGroup(
      key: runtime.decodeKey(data),
      policy: runtime.unpackOptionalField<FocusTraversalPolicy>(data, 'policy'),
      descendantsAreFocusable: runtime.unpackRequiredField<bool>(
        data,
        'descendantsAreFocusable',
      ),
      descendantsAreTraversable: runtime.unpackRequiredField<bool>(
        data,
        'descendantsAreTraversable',
      ),
      onFocusNodeCreated: runtime.unpackOptionalCallback(
        data,
        'onFocusNodeCreated',
      ),
      child: runtime.unpackRequiredField<Widget>(data, 'child'),
    );
  }
}

class FlutFocusTraversalOrder {
  FlutFocusTraversalOrder._();

  static FocusTraversalOrder? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return FocusTraversalOrder(
      key: runtime.decodeKey(data),
      order: runtime.unpackRequiredField<FocusOrder>(data, 'order'),
      child: runtime.unpackRequiredField<Widget>(data, 'child'),
    );
  }
}

File: flut/flutter/widgets/focus_traversal.py
from typing import override

from flut._flut.engine import (
    FlutAbstractObject,
    FlutEnumObject,
    FlutValueObject,
    _flut_pack_value,
)
from flut._flut.runtime import _flut_unpack_required_field
from flut.flutter.widgets.framework import Widget


class TraversalDirection(FlutEnumObject):
    up: "TraversalDirection"
    right: "TraversalDirection"
    down: "TraversalDirection"
    left: "TraversalDirection"


class TraversalEdgeBehavior(FlutEnumObject):
    closedLoop: "TraversalEdgeBehavior"
    leaveFlutterView: "TraversalEdgeBehavior"
    parentScope: "TraversalEdgeBehavior"
    stop: "TraversalEdgeBehavior"


class FocusTraversalPolicy(FlutAbstractObject):
    _flut_type = "FocusTraversalPolicy"

    def __init__(self, *, requestFocusCallback=None):
        super().__init__()
        self.requestFocusCallback = requestFocusCallback

    def sortDescendants(self, descendants, currentNode):
        raise NotImplementedError

    def invalidateScopeData(self, node):
        pass

    def changedScope(self, *, node=None, oldScope=None):
        pass

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        if self.requestFocusCallback is not None:
            result["requestFocusCallback"] = _flut_pack_value(self.requestFocusCallback)
        sort_callable = self._register_action(
            self.sortDescendants,
            base_class=FocusTraversalPolicy,
            method_name="sortDescendants",
        )
        if sort_callable is not None:
            result["sortDescendants"] = sort_callable._flut_pack()
        invalidate_callable = self._register_action(
            self.invalidateScopeData,
            base_class=FocusTraversalPolicy,
            method_name="invalidateScopeData",
        )
        if invalidate_callable is not None:
            result["invalidateScopeData"] = invalidate_callable._flut_pack()
        changed_callable = self._register_action(
            self.changedScope,
            base_class=FocusTraversalPolicy,
            method_name="changedScope",
        )
        if changed_callable is not None:
            result["changedScope"] = changed_callable._flut_pack()
        return result


class ReadingOrderTraversalPolicy(FocusTraversalPolicy):
    _flut_type = "ReadingOrderTraversalPolicy"

    def __init__(self, *, requestFocusCallback=None):
        super().__init__(requestFocusCallback=requestFocusCallback)


class OrderedTraversalPolicy(FocusTraversalPolicy):
    _flut_type = "OrderedTraversalPolicy"

    def __init__(self, *, secondary=None, requestFocusCallback=None):
        super().__init__(requestFocusCallback=requestFocusCallback)
        self.secondary = secondary

    @override
    def _flut_pack(self) -> dict:
        result = super()._flut_pack()
        if self.secondary is not None:
            result["secondary"] = _flut_pack_value(self.secondary)
        return result


class WidgetOrderTraversalPolicy(FocusTraversalPolicy):
    _flut_type = "WidgetOrderTraversalPolicy"

    def __init__(self, *, requestFocusCallback=None):
        super().__init__(requestFocusCallback=requestFocusCallback)


class FocusOrder(FlutValueObject):
    _flut_type = "FocusOrder"

    def __init__(self):
        super().__init__()


class NumericFocusOrder(FocusOrder):
    _flut_type = "NumericFocusOrder"

    def __init__(self, order: float):
        super().__init__()
        self.order = order

    @staticmethod
    def _flut_unpack(data: dict) -> "NumericFocusOrder":
        return NumericFocusOrder(
            order=_flut_unpack_required_field(data, "order"),
        )

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        result["order"] = _flut_pack_value(self.order)
        return result


class FocusTraversalGroup(Widget):
    _flut_type = "FocusTraversalGroup"

    def __init__(
        self,
        *,
        key=None,
        policy=None,
        descendantsAreFocusable: bool = True,
        descendantsAreTraversable: bool = True,
        onFocusNodeCreated=None,
        child: Widget,
    ):
        super().__init__(key=key)
        self.policy = policy
        self.descendantsAreFocusable = descendantsAreFocusable
        self.descendantsAreTraversable = descendantsAreTraversable
        self.onFocusNodeCreated = onFocusNodeCreated
        self.child = child

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        if self.policy is not None:
            result["policy"] = _flut_pack_value(self.policy)
        result["descendantsAreFocusable"] = _flut_pack_value(
            self.descendantsAreFocusable
        )
        result["descendantsAreTraversable"] = _flut_pack_value(
            self.descendantsAreTraversable
        )
        if self.onFocusNodeCreated is not None:
            result["onFocusNodeCreated"] = self._register_action(
                self.onFocusNodeCreated, "void Function(FocusNode)"
            )._flut_pack()
        result["child"] = _flut_pack_value(self.child)
        return result


class FocusTraversalOrder(Widget):
    _flut_type = "FocusTraversalOrder"

    def __init__(self, *, key=None, order, child: Widget):
        super().__init__(key=key)
        self.order = order
        self.child = child

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        result["order"] = _flut_pack_value(self.order)
        result["child"] = _flut_pack_value(self.child)
        return result
